### DISCUSSION TITLE
Both C++ and Python implementation of games necessary for some algorithms?

### DISCUSSION THREAD

**Discussion by regenbogenfrosch** [deepmind/open_spiel#975] [Q&A] [answered]
Hello all open_spiel developers

Thank you very much for releasing this wonderful project!

I am just new to RL and open_spiel in general and have some questions that might sound easy for you:
I am currently using open_spiel in my university work, and trying to understand RL for my project.
I have rewritten the "Tower Of Hanoi" problem as a k-general sum single player game in the C++ code part of open_spiel as described in the tutorial. And it works. Random plays run very well until the TerminalState is reached, at least for small problem sizes.
I also managed successfully to run the Q-Learning and SARSA learning algorithm (from one of the examples file adapted) and those algorithms also converge to the solution.

Now I want to test the "real" RL algorithms, as DQN, A2C and EVA. Unfortunately, I have found examples for using these algorithms only in the python directory of open_spiel.
Now, my silly question: does this mean, I have to rewrite my "myHanoi.cc" Code for my game completely new (and identical) in python (for example: myHanoi.py)? Or is it possible to use my written game and just load it from the C++ code directory somehow?
Sorry, for the silly sounding question, but I missed that somehow in the documentation.

Anyways, this is a cool framework, thanks again.

Best Regards
Frank

**Comment by jblespiau** [comment, marked as answer]
Hello,

games written in C++ are exposed to Python via Pybind11, so they can be used from Python without having to reimplement them.

This is documented here:
- https://openspiel.readthedocs.io/en/latest/developer_guide.html#c-and-python-implementations

It's also possible to add additional methods to a specific game, see https://github.com/deepmind/open_spiel/blob/master/open_spiel/python/pybind11/games_chess.cc for one example of additional methods on a game, which is called in https://github.com/deepmind/open_spiel/blob/master/open_spiel/python/pybind11/pyspiel.cc

**Comment by lanctot** [comment]
Sorry for the very late reply, I don't normally check this discussion area and there are no notification or anything int the main UI to indicate new ones. :(

What JB said is perfect (you only need to implement it once). The game gets "registered" into OpenSpiel so that `open_spiel::LoadGame` can find it (also exposed to python as `pyspiel.load_game`, see the example linked below).

 You can test loading your game from python using this example: https://github.com/deepmind/open_spiel/blob/master/open_spiel/python/examples/example.py (which echoes the same example in C++).
